compiler/src/iree/compiler/Dialect/Stream/Transforms/AutomaticReferenceCounting.cpp

```diff
@@ -0,0 +1,728 @@
+    }
+    Value baseResource =
+        IREE::Util::TiedOpInterface::findTiedBaseValue(resource);
+    aliases.insert({resource, baseResource});
+    return baseResource;
+  }
+};
+
+// Returns the last defined SSA value in the block in |timepoints|.
+// All timepoints must be in the same block.
+static Value getLastTimepointInBlock(TimepointSet &timepoints) {
+  if (timepoints.empty()) {
+    return nullptr;
+  } else if (timepoints.size() == 1) {
+    return *timepoints.begin();
+  }
+  Value lastTimepoint;
+  for (auto timepoint : timepoints) {
+    if (!lastTimepoint) {
+      lastTimepoint = timepoint;
+    } else {
+      auto *timepointOp = timepoint.getDefiningOp();
+      auto *lastTimepointOp = lastTimepoint.getDefiningOp();
+      if (!timepointOp) {
+        continue; // block arg
+      } else if (!lastTimepointOp) {
+        lastTimepoint = timepoint; // last found was a block arg, this isn't
+      } else if (lastTimepointOp->isBeforeInBlock(timepointOp)) {
+        lastTimepoint = timepoint;
+      }
+    }
+  }
+  return lastTimepoint;
+}
+
+// Returns a FusedLoc with the location of all |timepoints| and the base |loc|.
+static Location getFusedLocFromTimepoints(Location loc,
+                                          TimepointSet &timepoints) {
+  auto locs = llvm::map_to_vector(
+      timepoints, [](Value timepoint) { return timepoint.getLoc(); });
+  locs.push_back(loc);
+  return FusedLoc::get(loc.getContext(), locs);
+}
+
+//===----------------------------------------------------------------------===//
+// --iree-stream-automatic-reference-counting
+//===----------------------------------------------------------------------===//
+
+static StringRef getFuncName(FunctionOpInterface funcOp) {
+  if (isa<IREE::Util::InitializerOp>(funcOp)) {
+    return "(initializer)";
+  } else {
+    return funcOp.getName();
+  }
+}
+
+static void processFuncOp(FunctionOpInterface funcOp) {
+  // Today we bail on unstructured control flow. Eventually we want to be able
+  // to only support structured control flow in stream but today it can
+  // sometimes leak through depending on how frontends lower ops.
+  if (funcOp.getBlocks().size() != 1) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "[arc] skipping function @" << getFuncName(funcOp)
+               << " as it contains unstructured control flow (>1 blocks)\n");
+    return;
+  }
+
+  // NOTE: with the amount of IR this prints we must amortize the AsmState.
+  // Do not print SSA values/ops without using it.
+  std::unique_ptr<AsmState> asmState;
+  LLVM_DEBUG({
+    asmState = std::make_unique<AsmState>(
+        funcOp, OpPrintingFlags().elideLargeElementsAttrs());
+  });
+  LLVM_DEBUG(llvm::dbgs() << "\n[arc] processing function @"
+                          << getFuncName(funcOp) << ":\n");
+
+  Liveness liveness(funcOp);
+  DenseSet<Value> indeterminateResources;
+  DenseSet<Value> handledResources;
+  for (auto [blockIndex, block] : llvm::enumerate(funcOp.getBlocks())) {
```

Comment:
I think this loop might be confusing given that only single basic block functions are supported right now. Maybe just remove it?

Reply:
I hope to extend it to support multiple blocks in the future so having the placeholder is nice. If anyone ever writes a relooper (I'm still waiting and holding out hope :) such that we can guarantee that scf is the only thing used for control flow at this stage of the compiler we can remove it.